textworld/logic/__init__.py

```diff
@@ -1601,6 +1601,26 @@ def apply(self, action: Action) -> bool:
         else:
             return False
 
+    def apply_on_copy(self, action: Action) -> Optional["State"]:
+        """
+        Apply an action to a copy of this state.
+
+        Parameters
+        ----------
+        action :
+            The action to apply.
+
+        Returns
+        -------
+        The copied state after the action has been applied or `None` if action
+        wasn't applicable.
+        """
+        state = self.copy()
```

Comment:
Probably worth avoiding the copy if the action doesn't apply.  Like https://github.com/Microsoft/TextWorld/blob/4b51259e2e08a190ecc52d403384aaac457c1df8/textworld/generator/chaining.py#L432 (which could be changed to use this new method too)

Reply:
Makes sense. Will do.